### PR TITLE
Implement KMS client

### PR DESCRIPTION
**Pull request by GrahamCampbell**
This implements what is needed for to perform encryption & description, and also envelope encryption & decryption.

EDIT: Also added `ListAliases` - will be useful to me. ;)

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
It is unfortunate that the generated code takes control of the base64 decoding and encoding. I'd rather not risk leaking information, and take control of this myself. This would allow me to use timing safe functions, and not perform unnecessary decodes and encodes, leaving secret material all over the memory.

**Comment by jderusse** [comment]
Encoding/decoding is part of the AWS spec, and I think it should be done by the SDK. 

I get your concern about security.

What's about adding a new flag in our manifest.json file telling that this service (or this operation / or this argument...) is sensitive and should be zeroized and using timesafe method?

**Comment by GrahamCampbell** [comment]
I don't think the official AWS SDK actually does the encoding and decoding in it's implementation, but I'm not sure there - I might be looking in the wrong place - I didn't try executing the code. For my particular use case, I don't think timing safety of the base64 decoding and encoding will matter, and I'm not to sure if the issue with leaving secret material in memory can actually be fixed, without possibly changes to symfony's http client (and possibly to php's source code and libcurl!) https://github.com/symfony/http-client/blob/5.4/Response/CommonResponseTrait.php#L41-L76.

**Comment by GrahamCampbell** [comment]
This is probably low risk, but I find it curious that lambdas don't scrub memory on shutdown, and only on startup (https://docs.aws.amazon.com/lambda/latest/operatorguide/execution-environment.html). This at least stops other lambdas from booting and looking in memory for stuff that looks like a private key of another customer... conveniently alongside that would be key arn. As long as the underlying machines are only used to run lambda, there is low risk to others getting to the secrets, but there is still the risk of code bugs (e.g. heartbleed) in the software i am running on my lambdas from exposing secrets.

Anyway, I am probably overthinking this. I am just aware of how easy it is to get key management wrong. ;)

**Comment by jderusse** [comment]
> I don't think the official AWS SDK actually does the encoding and decoding in it's implementation

Actually, it does :wink: 

The `CiphertextBlob` property use a `blob` type (https://github.com/aws/aws-sdk-php/blob/fc4ac03c6ec4acea81c01e8b785c8606943d1fd6/src/data/kms/2014-11-01/api-2.json#L897-L901), and the SDK encode and decode such value
https://github.com/aws/aws-sdk-php/blob/fc4ac03c6ec4acea81c01e8b785c8606943d1fd6/src/Api/Serializer/JsonBody.php#L87
https://github.com/aws/aws-sdk-php/blob/fc4ac03c6ec4acea81c01e8b785c8606943d1fd6/src/Api/Parser/JsonParser.php#L64

The comments also state that the values is base64 encoded when using the API or CLI directly
```php
    /**
     * The encrypted plaintext. When you use the HTTP API or the Amazon Web Services CLI, the value is Base64-encoded.
     * Otherwise, it is not Base64-encoded.
     */
    private $ciphertextBlob;
```

note: The generated code with encoding/decoding is generic and also already used in several places in AsyncAws (where asking to the developer to take care of encode/decode wouldn't have made sense) ie: Sqs message attribute, Ses RAW body, Rds Value, ...

**Review by jderusse — APPROVED** [review]
Thank you for this addition. Just few comments, and I'll be happy to merge.

If needed, security hardening (time/memory constant encoding, zeroing memory) could be addressed in dedicated PR

File: src/Service/Kms/tests/Integration/KmsClientTest.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        $result->resolve();
+
+        // self::assertTODO(expected, $result->getCiphertextBlob());
+        // self::assertTODO(expected, $result->getPlaintext());
+        self::assertSame('changeIt', $result->getKeyId());
+    }
+
+    private function getClient(): KmsClient
+    {
+        self::fail('Not implemented');
```
[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply src/Service/Kms/tests/Integration/KmsClientTest.php:103]: Gonna try it on a live AWS account today.

[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply src/Service/Kms/tests/Integration/KmsClientTest.php:103]: `ListAliases` is working. Will test the others in a bit.

[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply src/Service/Kms/tests/Integration/KmsClientTest.php:103]: Tested the other methods too. Working. :shipit:

**Comment by GrahamCampbell** [comment]
I think this is ready to 🚢 now @jderusse. :)

**Comment by jderusse** [comment]
@GrahamCampbell I've added functional tests (with docker local-kms image) and added the operations createKey / createAlias to be able to seed the tests

**Comment by jderusse** [comment]
thank you

**Comment by GrahamCampbell** [comment]
🎉

**Pull request merged**